pkg/operator/cluster.go

```diff
@@ -0,0 +1,250 @@
+	}
+	return nil
+}
+
+func (t *clusterTPR) findAllClusters() error {
+	logger.Info("finding existing clusters...")
+	clusterList, err := t.getClusterList()
+	if err != nil {
+		return err
+	}
+	logger.Infof("found %d clusters", len(clusterList.Items))
+	for i := range clusterList.Items {
+		c := clusterList.Items[i]
+
+		ns := c.Spec.Namespace
+		existingCluster := cluster.New(c.Spec, t.context.factory, t.context.clientset)
+		t.tracker.add(ns, c.Metadata.ResourceVersion)
+		t.clusters[ns] = existingCluster
+
+		logger.Infof("resuming cluster %s in namespace %s", c.Metadata.Name, ns)
+		t.startCluster(existingCluster)
+	}
+
+	t.watchVersion = clusterList.Metadata.ResourceVersion
+	return nil
+}
+
+func (t *clusterTPR) startCluster(c *cluster.Cluster) {
+	if t.devicesInUse && c.Spec.UseAllDevices {
+		logger.Warningf("devices in more than one namespace not supported. ignoring devices in namespace %s", c.Spec.Namespace)
+		c.Spec.UseAllDevices = false
+	}
+
+	if c.Spec.UseAllDevices {
+		t.devicesInUse = true
+	}
+
+	go func() {
+		err := c.CreateInstance()
+		if err != nil {
+			logger.Errorf("failed to create cluster in namespace %s. %+v", c.Spec.Namespace, err)
+			return
+		}
+		c.Monitor(t.tracker.stopChMap[c.Spec.Namespace])
+	}()
+}
+
+func (t *clusterTPR) isClustersCacheStale(currentClusters []cluster.Cluster) bool {
+	if len(t.tracker.clusterRVs) != len(currentClusters) {
+		return true
+	}
+
+	for _, cc := range currentClusters {
+		rv, ok := t.tracker.clusterRVs[cc.Metadata.Name]
+		if !ok || rv != cc.Metadata.ResourceVersion {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (t *clusterTPR) getClusterList() (*cluster.ClusterList, error) {
+	b, err := getRawList(t.context, t)
+	if err != nil {
+		return nil, err
+	}
+
+	clusters := &cluster.ClusterList{}
+	if err := json.Unmarshal(b, clusters); err != nil {
+		return nil, err
+	}
+	return clusters, nil
+}
+
+func (t *clusterTPR) Watch() error {
+	logger.Infof("start watching %s tpr: %s", t.Name(), t.watchVersion)
+	defer t.tracker.stop()
+
+	eventCh, errCh := t.watch()
+
+	go func() {
+		timer := k8sutil.NewPanicTimer(
+			time.Minute,
+			fmt.Sprintf("unexpected long blocking (> 1 Minute) when handling %s event", t.Name()))
+
+		for event := range eventCh {
+			timer.Start()
+
+			c := event.Object
+
+			switch event.Type {
+			case kwatch.Added:
+				ns := c.Spec.Namespace
+				if ns == "" {
+					logger.Errorf("missing namespace attribute in rook spec")
+					continue
+				}
+
+				newCluster := cluster.New(c.Spec, t.context.factory, t.context.clientset)
+				t.tracker.add(ns, c.Metadata.ResourceVersion)
+				t.clusters[ns] = newCluster
+
+				logger.Infof("starting new cluster %s in namespace %s", c.Metadata.Name, ns)
+				t.startCluster(newCluster)
+
+			case kwatch.Modified:
+				logger.Infof("modifying a cluster not implemented")
+
+			case kwatch.Deleted:
+				logger.Infof("deleting a cluster not implemented")
+			}
+
+			timer.Stop()
+		}
+	}()
+	return <-errCh
```

Comment:
as discussed, if this error channel is not being used then it and the goroutine can likely be removed.